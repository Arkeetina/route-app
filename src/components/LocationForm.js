import React, {Component} from 'react';

export default class LocationForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      // drop and start off locations props check is set up for the tests
      startOffLocation: props.locations ? props.locations.startOffLocation : '',
      dropOffLocation: props.locations ? props.locations.dropOffLocation : '',
      error: '',
      disabled: true,
    }
  }

  //sets start off locations state
  onStartOffLocationChange = event => {
    const startOffLocation = event.target.value;
    this.setState(() => ({ startOffLocation }));
  }

  //sets drop off locations state
  onDropOffLocationChange = event => {
    const dropOffLocation = event.target.value;
    this.setState(() => ({ dropOffLocation }));
  }

  //submits form with start off and drop off locations
  onSubmit = event => {
    event.preventDefault();
    if (!this.state.startOffLocation.trim() || !this.state.dropOffLocation.trim()) {
      this.setState(() => ({ error: 'Please provide start and drop off locations' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.setState(() => ({ disabled: true }));
      let startOffLocation = this.state.startOffLocation;
      let dropOffLocation = this.state.dropOffLocation;
      this.props.onSubmit({
        startOffLocation,
        dropOffLocation,
      })
    }
  }

  // enables Send Location button on focus
  onFocus = () => {
    this.setState(() => ({ disabled: false }));
  }

  render() {
    const { startOffLocation, error, dropOffLocation, disabled } = this.state;

    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          onFocus={this.onFocus}
          placeholder="Enter start off location"
          type="text"
          value={this.state.startOffLocation}
          onChange={this.onStartOffLocationChange}
        />
        <input
          className="text-input"
          onFocus={this.onFocus}
          placeholder="Enter drop off location"
          type="text"
          value={this.state.dropOffLocation}
          onChange={this.onDropOffLocationChange}
        />
        <button
          disabled={!startOffLocation || !dropOffLocation || disabled}
          className="button"
          type="submit"
        >
          Send Location
        </button>
      </form>
    );
  }
}

