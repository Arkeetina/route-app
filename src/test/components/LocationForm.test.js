import React from 'react';
import { shallow } from 'enzyme';
import LocationForm from '../../components/LocationForm';
import locations from './fixtures/locations';

test('should correctly render LocationForm', () => {
  const wrapper = shallow(<LocationForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  expect(wrapper).toMatchSnapshot();
  const wrapper = shallow(<LocationForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });

  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set start off location on input change', () => {
  const value = 'Some destionation';
  const wrapper = shallow(<LocationForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value },
  });
  expect(wrapper.state('startOffLocation')).toBe(value);
});

test('should set drop off location on input change', () => {
  const value = 'Some destionation';
  const wrapper = shallow(<LocationForm />);
  wrapper.find('input').at(1).simulate('change', {
    target: { value },
  });
  expect(wrapper.state('dropOffLocation')).toBe(value);
});

test('should call onSubmit form on valid form submision', () => {
  const onSubmitSpy = jest.fn();
  onSubmitSpy(locations[0]);
  const wrapper = shallow(<LocationForm locations={locations[0]} onSubmit={onSubmitSpy} />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });

  expect(wrapper.state('error')).toBe('');
  expect(wrapper.state('disabled')).toBe(true);
  expect(onSubmitSpy).toHaveBeenLastCalledWith(locations[0]);
});

test('should set disabled to false on input focuse', () => {
  const wrapper = shallow(<LocationForm />);
  wrapper.find('input').at(0).simulate('focus');
  wrapper.find('input').at(1).simulate('focus');

  expect(wrapper.state('disabled')).toBe(false);
});

test('should set disabled to true for empty inputs', () => {
  const wrapper = shallow(<LocationForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: '',
  });
  wrapper.find('input').at(1).simulate('change', {
    target: '',
  });

  expect(wrapper.state('disabled')).toBe(true);
});

test('should set disabled to true for one empty input', () => {
  const value = 'Some destionation';
  const wrapper = shallow(<LocationForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: '',
  });
  wrapper.find('input').at(1).simulate('change', {
    target: { value },
  });

  expect(wrapper.state('disabled')).toBe(true);
});
