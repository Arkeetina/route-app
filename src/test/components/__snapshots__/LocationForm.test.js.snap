// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render LocationForm 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter start off location"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter drop off location"
    type="text"
    value=""
  />
  <button
    className="button"
    disabled={true}
    type="submit"
  >
    Send Location
  </button>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter start off location"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter drop off location"
    type="text"
    value=""
  />
  <button
    className="button"
    disabled={true}
    type="submit"
  >
    Send Location
  </button>
</form>
`;

exports[`should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide start and drop off locations
  </p>
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter start off location"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Enter drop off location"
    type="text"
    value=""
  />
  <button
    className="button"
    disabled={true}
    type="submit"
  >
    Send Location
  </button>
</form>
`;
